Add Modal beam example (#27327)
idaholab/moose: modules/solid_mechanics/doc/content/modules/solid_mechanics/1d_elastic_waves.md

In MOOSE, modal analysis is set up using the Eigenvalue Executioner. The
Eigenvalue System allows you to define which kernels go into which matrix. It's important
to note that in modal analysis, Neumann boundary conditions are
ignored since we are interested in the free vibration response of the structure.
Only Dirichlet boundary conditions are considered.

To demonstrate modal analysis in MOOSE, we will visit the cantilever beam example that will be discussed in more detail in the next section on frequency response functions. The goal here is to calculate the first few natural frequencies and visualize the corresponding mode shapes of the cantilever beam.
The strong form for the modal analysis of the cantilever beam is:
\begin{equation}
\nabla \cdot \boldsymbol{\sigma}(\mathbf{u}) = -\rho \omega^2 \mathbf{u}
\end{equation}
This equation represents the generalized eigenvalue problem for modal analysis.
The terms on the left side, which involve the stress divergence, will contribute
to the "A" matrix, and the terms on the right side, which are the mass terms, will contribute
to the "B" matrix.

## Cantilever Beam Example

> The cantilever beam shown in [cantilever] is subjected to a time harmonic force on the right side in the out-of-plane and vertical directions.  In this example the frequency of the time varying load is swept over a range.  The displacement at the midpoint of the beam is recorded at each frequency.  This type of output, displacement as a function of frequency, is a frequency response function (FRF) or transfer function.  Frequencies corresponding to the displacement peaks in the FRF indicate natural frequencies/modes.

Author: Different directions. This is a 3D problem.

The analytic first and second natural frequencies for this system bending in directions tangential to the beam axis are:\\
$\omega_{1a}=$509Hz\\
$\omega_{1b}=$763Hz\\
Both of these frequencies use the same value of $k_1$, but with the moment of inertia recomputed for bending about the different widths, where the lower frequency is the first bending mode about the thin direction ($h=a=$0.1m) and the next higher frequency is the first bending mode about the thicker direction ($h=b=$0.15m).

## Modal Analysis: Cantilever Beam Example

In the Kernels block, we define the kernels that contribute to the "A" and "B"
matrices in the eigenvalue problem. The StressDivergenceTensor kernels represent the stress divergence terms and contribute to the "A"
matrix. The ADCoefReaction kernels
represent the mass terms and contribute to the "B" matrix. The
[!param](/Kernels/CoefReaction/coefficient) in [kernels/CoefReaction.md] is set
to a negative value which corresponds to a positive density. The
`extra_vector_tags = 'eigen'` parameter is used to indicate that these kernels
contribute to "B" matrix. In other literature the A matrix would be the

> stiffness matrix (K) and B would be the Mass matrix(M).

Reviewer: ```suggestion
stiffness matrix (K) and B would be the Mass matrix (M).
```

[DirichletBC.md] is used, an [EigenDirichletBC.md] should be set also.

!listing test/tests/modal_analysis/modal.i block=BCs id=m_bc caption=BCs for
matrix A and B.

The Executioner block specifies the eigenvalue solver settings. The `type =
Eigenvalue` indicates that we are solving an eigenvalue problem. Since we are
solving a linear eigenvalue problem we can use a
[!param](/Executioner/Eigenvalue/solve_type) that can get multiple eigenvalues
at once, in this case that is `KRYLOVSCHUR`. The
[!param](/Executioner/Eigenvalue/which_eigen_pairs) parameter determines which
eigenvalues to compute and for modal analysis the smallest eigenvalues are usually of interest. The
[!param](/Executioner/Eigenvalue/n_eigen_pairs) parameter sets the number of eigenvalue pairs to compute.

!listing test/tests/modal_analysis/modal.i block=Executioner id=m_exec
caption=Eigenvalue Executioner

To output all the eigenvalues solved in the system we can use the
[vectorpostprocessors/Eigenvalues.md] vectorpostprocessor. While we have solved for the two smallest
eigenvalues, currently MOOSE only has the ability to output a single eigenvector. To
change which eigenvector is outputted adjust the index in

> [!param](/Problem/EigenProblem/active_eigen_index)and rerun the simulation.

Reviewer: ```suggestion
[!param](/Problem/EigenProblem/active_eigen_index) and rerun the simulation.
```

!listing test/tests/modal_analysis/modal.i block=VectorPostprocessors Problem
id=m_prob caption=Eigenvalues

For this coarse mesh example, we were able to determine the first two
$\omega$'s: 645.1 Hz and 855 Hz. Which are close to the theoretical result of
509 Hz and 763 Hz, and as the mesh is refined the results would converge on to the

> true solutions. The results are similar to the frequency response function as

Reviewer: did anyone check?
what refinement do you need?

Author: With an uniform refinement of 3, we get 513hz and 758hz. IIRC, since our poission's ratio isn't 0 it won't match perfectly. I added this to contrast the original method, so I didn't change any material properties or mesh resolution that was used.

# Frequency Domain Analysis

> For frequency domain analysis for Solid Mechanics MOOSE has two main methods
> modal analysis and frequency response functions. The following example will go

Reviewer 2: ```suggestion
The MOOSE Solid Mechanics module performs frequency domain analysis using modal analysis and frequency response functions. The following example will go
```

natural frequencies (eigenvalues) and mode shapes (eigenvectors) of a structure.
The natural frequencies represent the frequencies at which the structure tends
to vibrate when subjected to a disturbance, while the mode shapes describe the
corresponding deformed shape of the structure at each natural frequency. Modal
analysis provides valuable insights into the dynamic behavior of a structure and
is often used for design, optimization, and troubleshooting purposes.

The general eigenvalue problem that needs to be solved for modal analysis is given by:
\begin{equation}
(\mathbf{A} - \omega^2 \mathbf{B})\mathbf{u} = 0
\end{equation}
where $\mathbf{A}$ holds the "noneigen" kernels, $\mathbf{B}$ holds the "eigen" kernels, $\omega$ are the natural frequencies, and $\mathbf{u}$ are the mode shapes.
In MOOSE, modal analysis is set up using the Eigenvalue Executioner. The
Eigenvalue System allows you to define which kernels go into which matrix. It's important
to note that in modal analysis, Neumann boundary conditions are
ignored since we are interested in the free vibration response of the structure.
Only Dirichlet boundary conditions are considered.

> 

Reviewer 2: Will you move lines 136-153 here that describe the analytic solution.  Then your results will follow it.  The analytic results are $\omega_{1a}=$509Hz\\
$\omega_{1b}=$763Hz. Do you know if you are using the same coarse mesh as the coarse mesh FRF results below.  If so, natural frequencies are much better, you should mention that.

The strong form for the modal analysis of the cantilever beam is:
\begin{equation}
\nabla \cdot \boldsymbol{\sigma}(\mathbf{u}) = -\rho \omega^2 \mathbf{u}
\end{equation}
This equation represents the generalized eigenvalue problem for modal analysis.
The terms on the left side, which involve the stress divergence, will contribute
to the "A" matrix, and the terms on the right side, which are the mass terms, will contribute
to the "B" matrix.

## Modal Analysis: Cantilever Beam Example

In the Kernels block, we define the kernels that contribute to the "A" and "B"
matrices in the eigenvalue problem. The StressDivergenceTensor kernels represent the stress divergence terms and contribute to the "A"
matrix. The ADCoefReaction kernels
represent the mass terms and contribute to the "B" matrix. The
[!param](/Kernels/CoefReaction/coefficient) in [kernels/CoefReaction.md] is set
to a negative value which corresponds to a positive density. The
`extra_vector_tags = 'eigen'` parameter is used to indicate that these kernels
contribute to "B" matrix.

!listing test/tests/modal_analysis/modal.i block=Kernels id=m_kernel caption=
Kernels for A and B

> The BCs block defines the boundary conditions for the problem. In this example, we have Dirichlet boundary conditions applied to the left boundary for all displacement components. The [DirichletBC.md] is used for the "A" matrix, while the [EigenDirichletBC.md] is used for the "B" matrix.

Reviewer 2: Would the boundary conditions on the B matrix ever be different from the A matrix?  If not, mention something about that.  It seems like there should have a eigen tag for BCs too instead of a separate BC.

!listing test/tests/modal_analysis/modal.i block=BCs id=m_bc caption=BCs for
matrix A and B.

The Executioner block specifies the eigenvalue solver settings. The `type =
Eigenvalue` indicates that we are solving an eigenvalue problem. Since we are
solving a linear eigenvalue problem we can use a
[!param](/Executioner/Eigenvalue/solve_type) that can get multiple eigenvalues
at once, in this case that is `KRYLOVSCHUR`. The
[!param](/Executioner/Eigenvalue/which_eigen_pairs) parameter determines which
eigenvalues to compute and for modal analysis the smallest eigenvalues are usually of interest. The
[!param](/Executioner/Eigenvalue/n_eigen_pairs) parameter sets the number of eigenvalue pairs to compute.

!listing test/tests/modal_analysis/modal.i block=Executioner id=m_exec
caption=Eigenvalue Executioner

To output all the eigenvalues solved in the system we can use the

> [vectorpostprocessors/Eigenvalues.md] vectorpostprocessor. While we have solved for the two smallest
> eigenvalues, currently MOOSE only has the ability to output a single one. To
> change which eigenvector is outputted adjust the index in
> [!param](/Problem/EigenProblem/active_eigen_index).

Reviewer 2: ```suggestion
[vectorpostprocessors/Eigenvalues.md] vectorpostprocessor. While we have solved for the two smallest
eigenvalues, currently MOOSE only has the ability to output a single eigenvector. To
change which eigenvector is outputted adjust the index in
[!param](/Problem/EigenProblem/active_eigen_index) and rerun the simulation.
```

!listing test/tests/modal_analysis/modal.i block=VectorPostprocessors Problem
id=m_prob caption=Eigenvalues

For this coarse mesh example, we were able to determine the first two
$\omega$'s: 552.8 Hz and 789 Hz. Which are close to the theoretical result of

> 600 Hz and 800 Hz, and as the mesh is refined the results would converge on to the

Reviewer 2: ```suggestion
509 Hz and 763 Hz, and as the mesh is refined the results would converge on to the
```

> 600 Hz and 800 Hz, and as the mesh is refined the results would converge on to the

Reviewer 2: Say that this is closer than the natural frequencies determined using the same mesh with and the frequency response function in the next section.

Frequency domain analyses can be advantageous over its time domain counterpart in several cases, for example, when the frequency spectrum of a signal consists of a few frequencies, or, when it is needed to have a better control of the numerical dispersion.

## Problem Description

The equations of motion for a one dimensional isotropic elastic solid is given by the following partial differential equation:
\begin{equation}
    -E\frac{\partial^2 u}{\partial x^2}+\rho \frac{\partial^2 u}{\partial t^2}=0
\label{eq1}
\end{equation}
where $E$ is the modulus of elasticity, and $\rho$ is the density.
To convert to the frequency domain, we consider that a plane wave given by
\begin{equation}
  u(x,t)= B e^{i(\omega t - k x)}
  \label{eq2}
\end{equation}
is a solution to [eq1] where $B$ is in general a complex number depending on the boundary conditions, $k^2=\omega^2\rho/E$ is the wave number, $\omega=2 \pi f$ where $f$ is the frequency, and $i=\sqrt{-1}$.

> By assuming [eq2] is a a solution to [eq1] we can solve [eq1] in the frequency domain by taking a Fourier transform of $u(x,t)$ to get $U(x,\omega)$.  The frequency domain version of [eq1] is the Helmholtz equation given by

Reviewer 2: ```suggestion
By assuming [eq2] is a solution to [eq1] we can solve [eq1] in the frequency domain by taking a Fourier transform of $u(x,t)$ to get $U(x,\omega)$.  The frequency domain version of [eq1] is the Helmholtz equation given by
```
